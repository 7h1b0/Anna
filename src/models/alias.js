const mongoose = require('mongoose');
const Joi = require('joi');

const Schema = mongoose.Schema;

const alias = new Schema({
  name: { type: String, required: true, unique: true },
  description: { type: String, required: true },
  sceneId: { type: String, required: true },
  enabled: { type: Boolean, required: true },
});

alias.statics.validate = function validate(data, callback) {
  const pattern = {
    name: Joi.string()
      .regex(/^[a-z_]{5,}$/)
      .required(),
    description: Joi.string()
      .trim()
      .min(5)
      .required(),
    sceneId: Joi.string()
      .regex(/^[a-z0-9]{24}$/)
      .required(),
    enabled: Joi.boolean().required(),
  };
  Joi.validate(data, pattern, callback);
};

module.exports = mongoose.model('Alias', alias);
