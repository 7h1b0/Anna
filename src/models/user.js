const mongoose = require('mongoose');
const Ajv = require('ajv');
const userSchema = require('../schemas/user');

const Schema = mongoose.Schema;

const user = new Schema({
  username: { type: String, required: true, unique: true },
  password: { type: String, required: true },
  token: { type: String, required: true },
});

user.statics.validate = function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(userSchema, data);
};

module.exports = mongoose.model('User', user);
