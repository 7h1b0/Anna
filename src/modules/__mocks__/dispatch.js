const dispatch = jest.fn(async () => {});
export default dispatch;
