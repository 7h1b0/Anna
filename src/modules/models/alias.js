import knex from '../../knexClient';
import Ajv from 'ajv';
import aliasSchema from '../schemas/alias';

export const TABLE = 'alias';
export const COLUMNS = [
  { aliasId: 'alias_id' },
  { sceneId: 'scene_id' },
  'name',
  'description',
  'enabled',
  { createdAt: 'created_at' },
  { updatedAt: 'updated_at' },
  { createdBy: 'created_by' },
];

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export function findAll() {
  return knex(TABLE).select(COLUMNS);
}

export function findById(aliasId) {
  return knex(TABLE)
    .first(COLUMNS)
    .where('alias_id', aliasId);
}

export function save({ sceneId, name, description, userId, enabled = true }) {
  return knex(TABLE)
    .insert({
      scene_id: sceneId,
      description,
      name,
      enabled,
      created_by: userId,
    })
    .then(([aliasId]) => {
      return {
        name,
        description,
        enabled,
        sceneId,
        createdBy: userId,
      };
    });
}

export function remove(aliasId) {
  return knex(TABLE)
    .where('alias_id', aliasId)
    .del();
}

export function findByIdAndUpdate(aliasId, payload) {
  return knex(TABLE)
    .update({ ...payload })
    .where('alias_id', aliasId);
}
