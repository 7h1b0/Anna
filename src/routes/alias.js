const { dispatch, actions } = require('../utils/');
const Alias = require('../models/alias');

module.exports = app => {
  app
    .route('/api/alias')
    .get((req, res) => {
      Alias.find({})
        .then(alias => res.send(alias))
        .catch(err => res.status(500).send({ err }));
    })
    .post((req, res) => {
      const isValid = Alias.validate(req.body);
      if (!isValid) {
        res.sendStatus(400);
      } else {
        const { name, description, sceneId, enabled = true } = req.body;
        const newAlias = new Alias({ name, description, enabled, sceneId });

        newAlias
          .save()
          .then(alias => res.status(201).send(alias))
          .catch(err => res.status(500).send({ err }));
      }
    });

  app
    .route('/api/alias/:id_alias([0-9a-z]{24})')
    .get((req, res) => {
      Alias.findById(req.params.id_alias)
        .then(alias => {
          if (!alias) {
            res.sendStatus(404);
          } else {
            res.send(alias);
          }
        })
        .catch(err => res.status(500).send({ err }));
    })
    .put((req, res) => {
      const isValid = Alias.validate(req.body);
      if (!isValid) {
        res.sendStatus(400);
      } else {
        Alias.findByIdAndUpdate(req.params.id_alias, req.body, { new: true })
          .then(updatedAlias => {
            if (!updatedAlias) {
              res.sendStatus(404);
            } else {
              res.send(updatedAlias);
            }
          })
          .catch(err => res.status(500).send({ err }));
      }
    })
    .patch((req, res) => {
      const isBadRequest =
        !req.body || (req.body.enabled !== false && req.body.enabled !== true);

      if (isBadRequest) {
        res.sendStatus(400);
      } else {
        Alias.findByIdAndUpdate(
          req.params.id_alias,
          { enabled: req.body.enabled },
          { new: true },
        )
          .then(updatedAlias => {
            if (!updatedAlias) {
              res.sendStatus(404);
            } else {
              res.send(updatedAlias);
            }
          })
          .catch(err => res.status(500).send({ err }));
      }
    })
    .delete((req, res) => {
      Alias.findByIdAndRemove(req.params.id_alias)
        .then(alias => {
          if (!alias) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    });

  app.get('/api/alias/:name([_a-z]{5,})', (req, res) => {
    Alias.findOne({ name: req.params.name })
      .then(alias => {
        if (!alias) return res.sendStatus(404);
        if (alias.enabled !== true) return res.sendStatus(403);

        return dispatch(actions.callScene(alias.sceneId)).then(() => res.end());
      })
      .catch(err => res.status(500).send({ err }));
  });
};
