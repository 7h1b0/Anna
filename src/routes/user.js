const User = require('../modules/models/user');
const { cryptoUtil } = require('../modules/');

module.exports = app => {
  app.post('/register', (req, res) => {
    const isValid = User.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    const { username, password } = req.body;
    Promise.all([cryptoUtil.random(36), cryptoUtil.hash(password)])
      .then(([token, hashedPassword]) => {
        return User.save({
          username,
          password: hashedPassword,
          token,
        });
      })
      .then(newUser => {
        return res.status(201).json(newUser);
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  });

  app.post('/login', (req, res) => {
    const isValid = User.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    const { username, password } = req.body;
    User.findByUsername(username)
      .then(findUser => {
        if (!findUser) {
          res.sendStatus(403);
        } else {
          cryptoUtil
            .verify(password, findUser.password)
            .then(isValid => {
              if (!isValid) {
                throw new Error('Wrong password');
              }
              const copyUser = {
                _id: findUser.user_id,
                username: findUser.username,
                token: findUser.token,
              };

              res.json(copyUser);
            })
            .catch(() => res.sendStatus(403));
        }
      })
      .catch(err => res.status(500).send({ err }));
  });

  app.get('/api/users', (req, res) => {
    User.findAll()
      .then(users => res.send(users))
      .catch(err => res.status(500).send({ err }));
  });

  app
    .route('/api/users/:id_user([0-9]+)')
    .patch((req, res) => {
      const isValid = User.validate(req.body);
      if (!isValid) {
        return res.sendStatus(400);
      }

      const { password } = req.body;
      cryptoUtil
        .hash(password)
        .then(hashedPassword =>
          User.findByIdAndUpdate(req.params.id_user, {
            password: hashedPassword,
          }),
        )
        .then(rowsAffected => {
          if (rowsAffected < 1) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    })
    .delete((req, res) => {
      User.delete(req.params.id_user)
        .then(user => {
          if (!user) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    });
};
