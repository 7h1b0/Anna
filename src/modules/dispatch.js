import { HUE_LIGHT, DIO, SCENE } from './type';
import { findById as findSceneById } from './models/scene';
import { setLightState } from '../services/hueService';
import dioAdd from '../services/dioService';

export default function dispatch({ type, id, body }) {
  switch (type) {
    case HUE_LIGHT:
      return setLightState(id, body);

    case DIO:
      return dioAdd(id, body.on);

    case SCENE:
      return findSceneById(id).then(scene => {
        if (scene) {
          return Promise.all(
            scene.actions.map(({ type, targetId, body }) =>
              dispatch({ type, id: targetId, body }),
            ),
          );
        }
        throw new Error(`Cannot find scene ${id}`);
      });

    default:
      return Promise.reject(`Unknow type ${type}`);
  }
}
