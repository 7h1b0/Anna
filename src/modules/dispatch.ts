import TYPES from './type';
import { AnnaAction } from './actions';
import { findById as findSceneById } from './models/scene';
import { setLightState } from 'services/hueService';
import dioAdd from 'services/dioService';

export default async function dispatch(action: AnnaAction): Promise<void> {
  switch (action.type) {
    case TYPES.HUE_LIGHT:
      return setLightState(action.targetId, action.body);

    case TYPES.DIO:
      return dioAdd(action.targetId, action.body.on);

    case TYPES.SCENE:
      await findSceneById(action.targetId).then(scene => {
        if (scene) {
          return Promise.all(scene.actions.map(action => dispatch(action)));
        }
        throw new Error(`Cannot find scene ${action.targetId}`);
      });
  }
}
