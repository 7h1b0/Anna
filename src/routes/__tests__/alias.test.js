import request from 'supertest';
import knex from '../../knexClient';
import * as Alias from '../../modules/models/alias';
import * as User from '../../modules/models/user';
import app from '../../index.js';
import dispatch from '../../modules/dispatch';

jest.mock('../../modules/dispatch');

const initAlias = [
  {
    alias_id: 1,
    scene_id: 1,
    name: 'testone',
    description: 'test',
    enabled: true,
    created_by: 1,
    created_at: new Date('2018-01-01'),
    updated_at: new Date('2018-01-02'),
  },
  {
    alias_id: 2,
    scene_id: 1,
    name: 'testtwo',
    description: 'test',
    enabled: false,
    created_by: 1,
    created_at: new Date('2018-01-01'),
    updated_at: new Date('2018-01-02'),
  },
];

const user = {
  user_id: 1,
  username: 'test',
  password: '$2a$10$4ftuQxquI/5NR3POJy.2O.DmscxoSdCBzUvlnX2iXGMxtpqhd3w6O', // anna
  token: '8e6a76928f76d23665f78ff3688ca86422d5',
};

describe('Alias API', () => {
  beforeAll(async () => {
    await knex(Alias.TABLE).truncate();
    await knex(User.TABLE).insert(user);
  });

  beforeEach(async () => {
    await knex(Alias.TABLE).insert(initAlias);
  });

  afterEach(async () => {
    await knex(Alias.TABLE).truncate();
    dispatch.mockClear();
  });

  afterAll(async () => {
    await knex(User.TABLE).truncate();
    await knex.destroy();
  });

  describe('/api/alias', () => {
    describe('GET', () => {
      it('should retun all alias', async () => {
        const response = await request(app)
          .get('/api/alias')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/alias')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });
    });

    describe('POST', () => {
      it('should create a new alias', async () => {
        const response = await request(app)
          .post('/api/alias')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 1,
            name: 'test_post',
            description: 'test',
            enabled: false,
          });

        expect(response.status).toBe(201);
        expect(response.body).toMatchSnapshot();

        const alias = await knex(Alias.TABLE)
          .first('*')
          .where('alias_id', 3);
        expect(alias).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
        });
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .post('/api/alias')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should retun 400 if request is invalid', async () => {
        const response = await request(app)
          .post('/api/alias')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 1,
            name: 'test_post',
            description: 'test',
            fake: false,
          });

        expect(response.status).toBe(400);
      });
    });
  });

  describe('/api/alias/:id', () => {
    describe('GET', () => {
      it('should retun an alias', async () => {
        const response = await request(app)
          .get('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/alias/2')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it("should retun 404 when alias don't exist", async () => {
        const response = await request(app)
          .get('/api/alias/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });

    describe('PATCH', () => {
      it('should update an alias', async () => {
        const response = await request(app)
          .patch('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 2,
            name: 'test_update',
            description: 'test',
            enabled: true,
          });

        expect(response.status).toBe(204);

        const alias = await knex(Alias.TABLE)
          .first('*')
          .where('alias_id', 1);

        expect(alias).toMatchSnapshot({
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
        });
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should retun 400 when request is invalid', async () => {
        const response = await request(app)
          .patch('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 2,
            name: 'test_update',
            description: 'test',
            fake: true,
          });

        expect(response.status).toBe(400);
      });

      it("should retun 404 when alias don't exist", async () => {
        const response = await request(app)
          .patch('/api/alias/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 2,
            name: 'test_update',
            description: 'test',
            enabled: true,
          });

        expect(response.status).toBe(404);
      });
    });

    describe('DELETE', () => {
      it('should delete an alias', async () => {
        const response = await request(app)
          .delete('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(204);

        const alias = await knex(Alias.TABLE)
          .select('*')
          .where('alias_id', 1);

        expect(alias).toHaveLength(0);
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete('/api/alias/1')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it("should retun 404 when alias don't exist", async () => {
        const response = await request(app)
          .delete('/api/alias/23')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
      });
    });
  });

  describe('/api/alias/:id/:enabled', () => {
    it('should enable an alias', async () => {
      const response = await request(app)
        .get('/api/alias/2/enable')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(204);

      const alias = await knex(Alias.TABLE)
        .first('*')
        .where('alias_id', 2);

      expect(alias).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      });
    });

    it('should disable an alias', async () => {
      const response = await request(app)
        .get('/api/alias/1/disable')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(204);

      const alias = await knex(Alias.TABLE)
        .first('*')
        .where('alias_id', 1);

      expect(alias).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
      });
    });

    it('should retun 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get('/api/alias/1/disable')
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBe(401);
    });

    it("should retun 404 when alias don't exist", async () => {
      const response = await request(app)
        .delete('/api/alias/23/disable')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(404);
    });
  });

  describe('/api/alias/:id/action', () => {
    it('should call a scene', async () => {
      const response = await request(app)
        .get('/api/alias/1/action')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(200);
      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: 'SCENE',
        id: 1,
      });
    });

    it('should not call a scene if alias is disabled', async () => {
      const response = await request(app)
        .get('/api/alias/2/action')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(403);
      expect(dispatch).toHaveBeenCalledTimes(0);
    });

    it("should retun 404 when alias don't exist", async () => {
      const response = await request(app)
        .get('/api/alias/3/action')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(404);
    });
  });
});
