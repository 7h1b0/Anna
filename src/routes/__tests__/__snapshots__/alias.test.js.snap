// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alias API /api/alias GET should retun all alias 1`] = `
Array [
  Object {
    "aliasId": 1,
    "createdAt": "2018-01-01T00:00:00.000Z",
    "createdBy": 1,
    "description": "test",
    "enabled": true,
    "name": "testone",
    "sceneId": 1,
    "updatedAt": "2018-01-02T00:00:00.000Z",
  },
  Object {
    "aliasId": 2,
    "createdAt": "2018-01-01T00:00:00.000Z",
    "createdBy": 1,
    "description": "test",
    "enabled": false,
    "name": "testtwo",
    "sceneId": 1,
    "updatedAt": "2018-01-02T00:00:00.000Z",
  },
]
`;

exports[`Alias API /api/alias POST should create a new alias 1`] = `
Object {
  "createdBy": 1,
  "description": "test",
  "enabled": false,
  "name": "test_post",
  "sceneId": 1,
}
`;

exports[`Alias API /api/alias POST should create a new alias 2`] = `
Object {
  "alias_id": 3,
  "created_at": Any<Date>,
  "created_by": 1,
  "description": "test",
  "enabled": false,
  "name": "test_post",
  "scene_id": 1,
  "updated_at": Any<Date>,
}
`;

exports[`Alias API /api/alias/:id GET should retun an alias 1`] = `
Object {
  "aliasId": 1,
  "createdAt": "2018-01-01T00:00:00.000Z",
  "createdBy": 1,
  "description": "test",
  "enabled": true,
  "name": "testone",
  "sceneId": 1,
  "updatedAt": "2018-01-02T00:00:00.000Z",
}
`;

exports[`Alias API /api/alias/:id PATCH should update an alias 1`] = `
Object {
  "alias_id": 1,
  "created_at": Any<Date>,
  "created_by": 1,
  "description": "test",
  "enabled": true,
  "name": "test_update",
  "scene_id": 2,
  "updated_at": Any<Date>,
}
`;

exports[`Alias API /api/alias/:id/:enabled should disable an alias 1`] = `
Object {
  "alias_id": 1,
  "created_at": Any<Date>,
  "created_by": 1,
  "description": "test",
  "enabled": false,
  "name": "testone",
  "scene_id": 1,
  "updated_at": Any<Date>,
}
`;

exports[`Alias API /api/alias/:id/:enabled should enable an alias 1`] = `
Object {
  "alias_id": 2,
  "created_at": Any<Date>,
  "created_by": 1,
  "description": "test",
  "enabled": true,
  "name": "testtwo",
  "scene_id": 1,
  "updated_at": Any<Date>,
}
`;
